adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java

```diff
@@ -662,34 +663,34 @@ protected final void onUnsubscribe(
                     Tags.COMPONENT.set(commandContext.getCurrentSpan(), getTypeName());
                     final Sample timer = metrics.startTimer();
                     final Command command = commandContext.getCommand();
-                    if (command.isValid()) {
-                        getTenantConfiguration(sub.getTenant(), commandContext.getTracingContext())
-                                .compose(tenantObject -> checkMessageLimit(tenantObject, command.getPayloadSize()))
-                                .setHandler(result -> {
-                                    if (result.succeeded()) {
+                    final Future<TenantObject> tenantTracker = getTenantConfiguration(sub.getTenant(),
+                            commandContext.getTracingContext());
+
+                    tenantTracker.compose(tenantObject -> {
+                        if (!command.isValid()) {
+                            return Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_BAD_REQUEST,
+                                    "malformed command message"));
+                        } else {
+                            return checkMessageLimit(tenantObject, command.getPayloadSize())
+                                    .compose(success -> {
                                         addMicrometerSample(commandContext, timer);
-                                        onCommandReceived(mqttEndpoint, sub, commandContext, cmdHandler);
-                                    } else {
-                                        // issue credit so that application(s) can send the next command
-                                        commandContext.reject(getErrorCondition(result.cause()), 1);
-                                        metrics.reportCommand(
-                                                command.isOneWay() ? Direction.ONE_WAY : Direction.REQUEST,
-                                                sub.getTenant(),
-                                                ProcessingOutcome.from(result.cause()),
-                                                command.getPayloadSize(),
-                                                timer);
-                                    }
-                                });
-                    } else {
+                                        onCommandReceived(tenantTracker.result(), mqttEndpoint, sub, commandContext,
+                                                cmdHandler);
+                                        return Future.succeededFuture();
+                                    });
+                        }
+                    }).otherwise(failure -> {
```

Comment:
this could be modified to resemble the code used in the AMQP adapter a little more, i.e. using *setHandler* here with both a *succeeded* and a *failed* branch. WDYT?

Reply:
As above, I would keep the `otherwise` block.